src/migration.py

```diff
@@ -0,0 +1,290 @@
+        case types_.DocumentMeta:
+            assert isinstance(file_meta, types_.DocumentMeta)  # nosec
+            report = _migrate_document(
+                document_meta=file_meta, discourse=discourse, docs_path=docs_path
+            )
+        case types_.IndexDocumentMeta:
+            assert isinstance(file_meta, types_.IndexDocumentMeta)  # nosec
+            report = _migrate_index(index_meta=file_meta, docs_path=docs_path)
+        # Edge case that should not be possible.
+        case _:  # pragma: no cover
+            raise exceptions.MigrationError(
+                f"internal error, no implementation for migration file, {file_meta=!r}"
+            )
+
+    logging.info("report: %s", report)
+    return report
+
+
+def _calculate_file_name(current_directory: Path, table_path: types_.TablePath) -> str:
+    """Calculate file name given table path from the index file and current path \
+        relative to the docs directory.
+
+    Args:
+        current_directory: current directory of the file relative to the docs directory.
+        table_path: table path of the file from the index file, of format path-to-file-filename.
+
+    Returns:
+        The filename derived by removing the directory path from given table path of the file.
+    """
+    return table_path.removeprefix(f"{calculate_table_path(current_directory)}-")
+
+
+def _extract_docs_from_table_rows(
+    table_rows: typing.Iterable[types_.TableRow],
+) -> typing.Iterable[types_.MigrationFileMeta]:
+    """Extract necessary migration documents to build docs directory from server.
+
+    Algorithm:
+        1.  For each table row:
+            1.1. If row level is smaller than current working level:
+                1.1.1. Yield GitkeepMeta if last working directory was empty.
+                1.1.2. Navigate to parent directory based on current level and row level.
+            1.2. If row is a directory:
+                1.2.1. Create a virtual directory with given path
+                1.2.2. Set created virtual directory as working directory.
+            1.3. If row is a file: Yield DocumentMeta
+        2. If last table row was a directory and yielded no DocumentMeta, yield GitkeepMeta.
+
+    Args:
+        table_rows: Table rows from the index file in the order of directory hierarchy.
+
+    Returns:
+        Migration documents with navlink to content.\
+            .gitkeep file with no content if empty directory.
+    """
+    table_rows = list(table_rows)
+    _validate_row_levels(table_rows=table_rows)
+
+    level = 0
+    last_dir_has_file = True  # Assume root dir is not empty.
+    last_dir_row: types_.TableRow | None = None
+    cwd = Path()
+    for row in table_rows:
+        # Next set of hierarchies, change cwd path
+        if row.level <= level:
+            if not last_dir_has_file and last_dir_row is not None:
+                yield types_.GitkeepMeta(path=cwd / GITKEEP_FILE, table_row=last_dir_row)
```

Comment:
This does handle the case since `last_dir_has_file` still evaluates to `False` and `last_dir_row` evaluates to `True` and therefore generates the `.gitkeep` files.
However, since this part of the code isn't too readable, I've refactored it with additional comments. Hopefully they are more readable!
Thank you for the feedback!